File: src/main/scala/codechicken/microblock/MicroRecipe.scala
package codechicken.microblock

import codechicken.microblock.MicroMaterialRegistry._
import codechicken.microblock.MicroRecipe._
import codechicken.microblock.api.{BlockMicroMaterial, MicroMaterial}
import codechicken.microblock.handler.MicroblockModContent
import codechicken.microblock.handler.MicroblockModContent._
import net.minecraft.block.Blocks
import net.minecraft.inventory.CraftingInventory
import net.minecraft.item.ItemStack
import net.minecraft.item.crafting.SpecialRecipe
import net.minecraft.util.ResourceLocation
import net.minecraft.world.World

class MicroRecipe(id: ResourceLocation) extends SpecialRecipe(id) {

    override def getSerializer = MicroblockModContent.microRecipeSerializer

    override def getResultItem = ItemMicroBlock.create(1, 1, BlockMicroMaterial.makeMaterialKey(Blocks.STONE.defaultBlockState))

    override def matches(icraft: CraftingInventory, world: World) = !assemble(icraft).isEmpty

    override def canCraftInDimensions(width: Int, height: Int): Boolean = width >= 3 && height >= 3

    override def assemble(icraft: CraftingInventory): ItemStack = {
        var res = getHollowResult(icraft)
        if (!res.isEmpty) return res
        res = getGluingResult(icraft)
        if (!res.isEmpty) return res
        res = getThinningResult(icraft)
        if (!res.isEmpty) return res
        res = getSplittingResult(icraft)
        if (!res.isEmpty) return res
        res = getHollowFillResult(icraft)
        res
    }

    def getHollowResult(icraft: CraftingInventory): ItemStack = {
        if (!getStackRowCol(icraft, 1, 1).isEmpty) return ItemStack.EMPTY

        val first = getStackRowCol(icraft, 0, 0)
        val factory = microFactory(first)
        val size = microSize(first)
        val material = microMaterial(first)
        if (first.isEmpty || first.getItem != itemMicroBlock || factory != 0) return ItemStack.EMPTY

        for (i <- 1 to 8 if i != 4) {
            val item = icraft.getItem(i)
            if (item.isEmpty || item.getItem != itemMicroBlock ||
                microMaterial(item) != material || microFactory(item) != factory || microSize(item) != size) {
                return ItemStack.EMPTY
            }
        }
        return create(8, 1, size, material)
    }

    def getGluingResult(icraft: CraftingInventory): ItemStack = {
        var size = 0
        var count = 0
        var smallest = 0
        var mcrFactory = 0
        var material: MicroMaterial = null
        for (i <- 0 until 9) {
            val item = icraft.getItem(i)
            if (!item.isEmpty) {
                if (item.getItem != itemMicroBlock) return ItemStack.EMPTY
                if (count == 0) {
                    size = microSize(item)
                    mcrFactory = microFactory(item)
                    material = microMaterial(item)
                    count = 1
                    smallest = size
                }
                else if (microFactory(item) != mcrFactory || microMaterial(item) != material) {
                    return ItemStack.EMPTY
                } else if (mcrFactory >= 2 && microSize(item) != smallest) {
                    return ItemStack.EMPTY
                } else {
                    smallest = Math.min(smallest, microSize(item))
                    count += 1
                    size += microSize(item)
                }
            }
        }

        if (count <= 1) return ItemStack.EMPTY

        mcrFactory match {
            case 3 => count match {
                case 2 => create(1, 0, smallest, material)
                case _ => ItemStack.EMPTY
            }
            case 2 => count match {
                case 2 => create(1, 3, smallest, material)
                case 4 => create(1, 0, smallest, material)
                case _ => ItemStack.EMPTY
            }
            case 1 | 0 =>
                val base = Seq(1, 2, 4).find(s => (s & size) != 0)
                if (base.isEmpty) {
                    create(size / 8, 0, 8, material)
                } else if (base.get <= smallest) {
                    ItemStack.EMPTY
                } else {
                    create(size / base.get, mcrFactory, base.get, material)
                }
            case _ => ItemStack.EMPTY
        }
    }

    def getSaw(icraft: CraftingInventory): (Saw, Int, Int) = {
        for (r <- 0 until 3)
            for (c <- 0 until 3) {
                val item = getStackRowCol(icraft, c, r)
                if (!item.isEmpty && item.getItem.isInstanceOf[Saw]) {
                    return (item.getItem.asInstanceOf[Saw], r, c)
                }
            }
        return (null, 0, 0)
    }

    def getThinningResult(icraft: CraftingInventory): ItemStack = {
        val (saw, row, col) = getSaw(icraft)
        if (saw == null) {
            return ItemStack.EMPTY
        }

        val item = getStackRowCol(icraft, col, row + 1)
        if (item.isEmpty) {
            return ItemStack.EMPTY
        }

        val size = microSize(item)
        val material = microMaterial(item)
        val mcrClass = microFactory(item)
        if (size == 1 || material == null || !canCut(saw, getStackRowCol(icraft, col, row), material)) {
            return ItemStack.EMPTY
        }

        for (r <- 0 until 3)
            for (c <- 0 until 3)
                if ((c != col || r != row && r != row + 1) &&
                    !getStackRowCol(icraft, c, r).isEmpty) {
                    return ItemStack.EMPTY
                }

        return create(2, mcrClass, size / 2, material)
    }

    val splitMap = Map(0 -> 3, 1 -> 3, 3 -> 2)

    def getSplittingResult(icraft: CraftingInventory): ItemStack = {
        val (saw, row, col) = getSaw(icraft)
        if (saw == null) return ItemStack.EMPTY
        val item = getStackRowCol(icraft, col + 1, row)
        if (item.isEmpty || item.getItem != itemMicroBlock) return ItemStack.EMPTY
        val mcrClass = microFactory(item)
        val material = microMaterial(item)
        if (!canCut(saw, getStackRowCol(icraft, col, row), material)) return ItemStack.EMPTY
        val split = splitMap.get(mcrClass)
        if (split.isEmpty) return ItemStack.EMPTY

        for (r <- 0 until 3)
            for (c <- 0 until 3)
                if ((r != row || c != col && c != col + 1) &&
                    !getStackRowCol(icraft, c, r).isEmpty) {
                    return ItemStack.EMPTY
                }

        return create(2, split.get, microSize(item), material)
    }

    def getHollowFillResult(icraft: CraftingInventory): ItemStack = {
        var cover: ItemStack = ItemStack.EMPTY
        for (i <- 0 until 9) {
            val item = icraft.getItem(i)
            if (!item.isEmpty) {
                if (item.getItem != itemMicroBlock || !cover.isEmpty || microFactory(item) != 1) {
                    return ItemStack.EMPTY
                } else {
                    cover = item
                }
            }
        }
        if (cover.isEmpty) return ItemStack.EMPTY
        return create(1, 0, microSize(cover), microMaterial(cover))
    }
}

object MicroRecipe {
    def create(amount: Int, factoryID: Int, size: Int, material: MicroMaterial): ItemStack = {
        if (size == 8) {
            val item = material.getItem.copy
            item.setCount(amount)
            return item
        }
        ItemMicroBlock.createStack(amount, factoryID, size, material)
    }

    def canCut(saw: Saw, sawItem: ItemStack, material: MicroMaterial): Boolean = {
        val sawStrength = saw.getCuttingStrength(sawItem)
        val matStrength = material.getCutterStrength
        sawStrength >= matStrength || sawStrength == MicroMaterialRegistry.getMaxCuttingStrength
    }

    def getStackRowCol(inv: CraftingInventory, row: Int, col: Int): ItemStack = inv.getItem(row + col * inv.getWidth)
}

File: src/main/java/codechicken/microblock/recipe/MicroRecipe.java
package codechicken.microblock.recipe;

import codechicken.lib.util.ItemUtils;
import codechicken.microblock.api.MicroMaterial;
import codechicken.microblock.init.CBMicroblockModContent;
import codechicken.microblock.init.CBMicroblockTags;
import codechicken.microblock.item.ItemMicroBlock;
import codechicken.microblock.item.MicroMaterialComponent;
import codechicken.microblock.util.MicroMaterialRegistry;
import net.minecraft.core.HolderLookup;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.crafting.CraftingBookCategory;
import net.minecraft.world.item.crafting.CraftingInput;
import net.minecraft.world.item.crafting.CustomRecipe;
import net.minecraft.world.item.crafting.RecipeSerializer;
import net.minecraft.world.level.Level;
import org.jetbrains.annotations.Nullable;

import static codechicken.microblock.init.CBMicroblockModContent.MICRO_BLOCK_ITEM;

/**
 * Created by covers1624 on 22/10/22.
 */
public class MicroRecipe extends CustomRecipe {

    private static final int[] splitMap = { 3, 3, -1, 2 };

    public MicroRecipe() {
        super(CraftingBookCategory.MISC);
    }

    @Override
    public RecipeSerializer<?> getSerializer() {
        return CBMicroblockModContent.MICRO_RECIPE_SERIALIZER.get();
    }

    @Override
    public boolean matches(CraftingInput input, Level level) {
        return !getAssemblyResult(input).isEmpty();
    }

    @Override
    public boolean canCraftInDimensions(int width, int height) {
        return width >= 3 && height >= 3;
    }

    @Override
    public ItemStack assemble(CraftingInput inv, HolderLookup.Provider registries) {
        return getAssemblyResult(inv);
    }

    private ItemStack getAssemblyResult(CraftingInput inv) {
        var res = getHollowResult(inv);
        if (!res.isEmpty()) return res;

        res = getGluingResult(inv);
        if (!res.isEmpty()) return res;

        res = getThinningResult(inv);
        if (!res.isEmpty()) return res;

        res = getSplittingResult(inv);
        if (!res.isEmpty()) return res;

        return getHollowFillResult(inv);
    }

    private static ItemStack getHollowResult(CraftingInput inv) {
        if (inv.width() != 3 || inv.height() != 3) return ItemStack.EMPTY;

        if (!inv.getItem(1, 1).isEmpty()) return ItemStack.EMPTY;

        ItemStack first = inv.getItem(0, 0);
        if (!first.is(MICRO_BLOCK_ITEM.get())) return ItemStack.EMPTY;

        MicroMaterialComponent materialComponent = MicroMaterialComponent.getComponent(first);
        if (materialComponent == null) return ItemStack.EMPTY;

        for (int x = 0; x < 3; x++) {
            for (int y = 0; y < 3; y++) {
                if (x == 0 && y == 0 || x == 1 && y == 1) continue;

                ItemStack stack = inv.getItem(x, y);
                if (!stack.is(MICRO_BLOCK_ITEM.get())) return ItemStack.EMPTY;

                MicroMaterialComponent comp = MicroMaterialComponent.getComponent(stack);
                if (comp == null) return ItemStack.EMPTY;

                if (!comp.equals(materialComponent)) return ItemStack.EMPTY;
            }
        }

        return create(8, 1, materialComponent.size(), materialComponent.material());
    }

    private static ItemStack getGluingResult(CraftingInput inv) {
        int size = 0;
        int count = 0;
        int smallest = 0;
        int mcrFactory = 0;
        MicroMaterial material = null;
        for (ItemStack stack : inv.items()) {
            if (!stack.isEmpty()) {
                if (!stack.is(MICRO_BLOCK_ITEM.get())) return ItemStack.EMPTY;

                MicroMaterialComponent component = MicroMaterialComponent.getComponent(stack);
                if (component == null) return ItemStack.EMPTY;

                if (count == 0) {
                    size = component.size();
                    mcrFactory = component.factoryId();
                    material = component.material();
                    count = 1;
                    smallest = size;
                    continue;
                }

                if (component.factoryId() != mcrFactory || component.material() != material) return ItemStack.EMPTY;
                if (mcrFactory >= 2 && component.size() != smallest) return ItemStack.EMPTY;

                smallest = Math.min(smallest, component.size());
                count += 1;
                size += component.size();
            }
        }

        if (material == null) return ItemStack.EMPTY;
        if (count <= 1) return ItemStack.EMPTY;

        return switch (mcrFactory) {
            case 3 -> count == 2 ? create(1, 0, smallest, material) : ItemStack.EMPTY;
            case 2 -> switch (count) {
                case 2 -> create(1, 3, smallest, material);
                case 3 -> create(1, 0, smallest, material);
                default -> ItemStack.EMPTY;
            };
            case 1, 0 -> {
                int base = -1;
                for (int s : new int[] { 1, 2, 4 }) {
                    if ((s & size) != 0) {
                        base = s;
                        break;
                    }
                }

                if (base == -1) yield create(size / 8, 0, 8, material);
                if (base <= smallest) yield ItemStack.EMPTY;
                yield create(size / base, mcrFactory, base, material);
            }
            default -> ItemStack.EMPTY;
        };
    }

    private static ItemStack getThinningResult(CraftingInput inv) {
        if (inv.items().size() != 2 || inv.width() != 2) return ItemStack.EMPTY;

        SawResult saw = findSaw(inv);
        if (saw == null) return ItemStack.EMPTY;

        ItemStack item = inv.getItem(saw.x == 0 ? 1 : 0, saw.y);
        if (item.isEmpty()) return ItemStack.EMPTY;

        int factory;
        int size;
        MicroMaterial material;

        MicroMaterialComponent comp = MicroMaterialComponent.getComponent(item);
        if (comp != null) {
            factory = comp.factoryId();
            size = comp.size();
            material = comp.material();
        } else {
            factory = 0;
            size = 8;
            material = findMaterial(item);
        }

        if (size == 1 || material == null) return ItemStack.EMPTY;

        return create(2, factory, size / 2, material);
    }

    private static ItemStack getSplittingResult(CraftingInput inv) {
        if (inv.items().size() != 2 || inv.height() != 2) return ItemStack.EMPTY;

        SawResult saw = findSaw(inv);
        if (saw == null) return ItemStack.EMPTY;

        ItemStack item = inv.getItem(saw.x, saw.y == 0 ? 1 : 0);
        if (!item.is(MICRO_BLOCK_ITEM)) return ItemStack.EMPTY;

        MicroMaterialComponent comp = MicroMaterialComponent.getComponent(item);
        if (comp == null) return ItemStack.EMPTY;

        int split = splitMap[comp.factoryId()];
        if (split == -1) return ItemStack.EMPTY;

        return create(2, split, comp.size(), comp.material());
    }

    private static ItemStack getHollowFillResult(CraftingInput inv) {
        if (inv.items().size() != 1) return ItemStack.EMPTY;

        ItemStack only = inv.items().getFirst();
        if (!only.is(CBMicroblockModContent.MICRO_BLOCK_ITEM)) return ItemStack.EMPTY;

        MicroMaterialComponent component = MicroMaterialComponent.getComponent(only);
        if (component == null || component.factoryId() != 1) return ItemStack.EMPTY;

        return create(1, 0, component.size(), component.material());
    }

    private static @Nullable SawResult findSaw(CraftingInput inv) {
        for (int x = 0; x < inv.width(); x++) {
            for (int y = 0; y < inv.height(); y++) {
                ItemStack item = inv.getItem(x, y);
                if (item.is(CBMicroblockTags.Items.TOOL_SAW)) {
                    return new SawResult(item, x, y);
                }
            }
        }
        return null;
    }

    private static ItemStack create(int amount, int factoryId, int size, MicroMaterial material) {
        if (size == 8) {
            return ItemUtils.copyStack(material.getItem(), amount);
        }
        return ItemMicroBlock.createStack(amount, factoryId, size, material);
    }

    @Nullable
    public static MicroMaterial findMaterial(ItemStack stack) {
        if (stack.isEmpty()) return null;
        for (MicroMaterial material : MicroMaterialRegistry.microMaterials()) {
            ItemStack mStack = material.getItem();
            if (ItemStack.isSameItemSameComponents(mStack, stack)) {
                return material;
            }
        }
        return null;
    }

    private record SawResult(ItemStack stack, int x, int y) { }
}
